File: src/main/java/net/snipsniper/utils/DrawUtils.java
package net.snipsniper.utils;

import java.awt.*;
import java.awt.geom.Rectangle2D;

public class DrawUtils {
    public static void drawRect(Graphics g, Rectangle rect) {
        if(rect != null)
            g.drawRect(rect.x, rect.y, rect.width, rect.height);
    }

    public static void fillRect(Graphics g, Rectangle rect) {
        if(rect != null)
            g.fillRect(rect.x, rect.y, rect.width, rect.height);
    }

    public static void drawRect(Graphics2D g, Rectangle rect) {
        if(rect != null)
            g.drawRect(rect.x, rect.y, rect.width, rect.height);
    }

    public static void fillRect(Graphics2D g, Rectangle rect) {
        if(rect != null)
            g.fillRect(rect.x, rect.y, rect.width, rect.height);
    }

    public static int pickOptimalFontSize(Graphics2D g, String title, int width, int height) {
        Rectangle2D rect;

        int fontSize = height; //initial value
        do {
            fontSize--;
            Font font = new Font("Arial", Font.PLAIN, fontSize);
            rect = getStringBoundsRectangle2D(g, title, font);
        } while (rect.getWidth() >= width || rect.getHeight() >= height);

        return fontSize;
    }

    public static Rectangle2D getStringBoundsRectangle2D(Graphics g, String title, Font font) {
        g.setFont(font);
        FontMetrics fm = g.getFontMetrics();
        return fm.getStringBounds(title, g);
    }

    public static void drawCenteredString(Graphics g, String text, Rectangle rect, Font font ) {
        FontMetrics metrics = g.getFontMetrics(font);
        int x = rect.x + (rect.width - metrics.stringWidth(text)) / 2;
        int y = rect.y + ((rect.height - metrics.getHeight()) / 2) + metrics.getAscent();
        g.setFont(font);
        g.drawString(text, x, y);
    }
}

File: src/main/kotlin/net/snipsniper/utils/DrawUtils.kt
package net.snipsniper.utils

import java.awt.Font
import java.awt.Graphics
import java.awt.Graphics2D
import java.awt.Rectangle
import java.awt.geom.Rectangle2D


class DrawUtils {
    companion object {
        fun drawRect(g: Graphics, rect: Rectangle) = g.drawRect(rect.x, rect.y, rect.width, rect.height)
        fun fillRect(g: Graphics, rect: Rectangle) = g.fillRect(rect.x, rect.y, rect.width, rect.height)
        fun drawRect(g: Graphics2D, rect: Rectangle) = g.drawRect(rect.x, rect.y, rect.width, rect.height)
        fun fillRect(g: Graphics2D, rect: Rectangle) = g.fillRect(rect.x, rect.y, rect.width, rect.height)

        fun pickOptimalFontSize(g: Graphics2D, title: String, width: Int, height: Int): Int {
            var rect: Rectangle2D
            var fontSize = height
            do {
                fontSize--
                Font("Arial", Font.PLAIN, fontSize).also {
                    rect = getStringBoundsRectangle2D(g, title, it)
                }
            } while (rect.width >= width || rect.height >= height)
            return fontSize
        }

        fun getStringBoundsRectangle2D(g: Graphics, title: String, font: Font): Rectangle2D {
            g.font = font
            return g.fontMetrics.getStringBounds(title, g)
        }

        fun drawCenteredString(g: Graphics, text: String, rect: Rectangle, font: Font) {
            val metrics = g.getFontMetrics(font)
            val x = rect.x + (rect.width - metrics.stringWidth(text)) / 2
            val y = rect.y + ((rect.height - metrics.height) / 2) + metrics.ascent
            g.font = font
            g.drawString(text, x, y)
        }
    }
}